diff --git a/kernel/irq/irqdomain.c b/kernel/irq/irqdomain.c
index 54a4d52..a76a66b 100644
--- a/kernel/irq/irqdomain.c
+++ b/kernel/irq/irqdomain.c
@@ -134,6 +134,7 @@ static unsigned int irq_domain_legacy_revmap(struct irq_domain *domain,
 	irq_hw_number_t first_hwirq = domain->revmap_data.legacy.first_hwirq;
 	int size = domain->revmap_data.legacy.size;
 
+pr_debug("%s(domain=%p, hwirq=%lu): first_hwirq=%lu, size=%d\n", __func__, domain, hwirq, first_hwirq, size);
 	if (WARN_ON(hwirq < first_hwirq || hwirq >= first_hwirq + size))
 		return 0;
 	return hwirq - first_hwirq + domain->revmap_data.legacy.first_irq;
@@ -221,6 +222,7 @@ struct irq_domain *irq_domain_add_legacy(struct device_node *of_node,
 	struct irq_domain *domain;
 	unsigned int i;
 
+pr_debug("%s(of_node=%p, size=%d, first_irq=0x%x, ops=%p, host_data=%p)\n", __func__, of_node, size, first_irq, ops, host_data);
 	domain = irq_domain_alloc(of_node, IRQ_DOMAIN_MAP_LEGACY, ops, host_data);
 	if (!domain)
 		return NULL;
diff --git a/kernel/irq/manage.c b/kernel/irq/manage.c
index dc4db32..165907a 100644
--- a/kernel/irq/manage.c
+++ b/kernel/irq/manage.c
@@ -1400,6 +1400,7 @@ int request_threaded_irq(unsigned int irq, irq_handler_t handler,
 	struct irq_desc *desc;
 	int retval;
 
+pr_debug("%s(irq=%u, devname=%s, dev_id=%p\n", __func__, irq, devname, dev_id);
 	/*
 	 * Sanity-check: shared interrupts must pass in a real dev-ID,
 	 * otherwise we'll have trouble later trying to figure out
diff --git a/kernel/resource.c b/kernel/resource.c
index d738698..831eea3 100644
--- a/kernel/resource.c
+++ b/kernel/resource.c
@@ -943,6 +943,7 @@ struct resource * __request_region(struct resource *parent,
 	DECLARE_WAITQUEUE(wait, current);
 	struct resource *res = alloc_resource(GFP_KERNEL);
 
+pr_debug("__request_region(parent=%p, start=0x%x, n=0x%x, name=%s, flags=0x%x): res=%p\n", parent, start, n, name, flags, res);
 	if (!res)
 		return NULL;
 
