diff --git a/drivers/i2c/busses/Kconfig b/drivers/i2c/busses/Kconfig
index 9ee5811..8f075d0 100644
--- a/drivers/i2c/busses/Kconfig
+++ b/drivers/i2c/busses/Kconfig
@@ -334,7 +334,7 @@ config I2C_AU1550
 
 config I2C_BCM2835
 	tristate "Broadcom BCM2835 I2C controller"
-	depends on ARCH_BCM2835
+	depends on (ARCH_BCM2835 || MACH_BCM2708)
 	help
 	  If you say yes to this option, support will be included for the
 	  BCM2835 I2C controller.
diff --git a/drivers/spi/Kconfig b/drivers/spi/Kconfig
index f0a2a9f..615f471 100644
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -77,7 +77,7 @@ config SPI_ATMEL
 
 config SPI_BCM2835
 	tristate "BCM2835 SPI controller"
-	depends on ARCH_BCM2835
+	depends on (ARCH_BCM2835 || MACH_BCM2708)
 	help
 	  This selects a driver for the Broadcom BCM2835 SPI master.
 
