diff --git a/drivers/video/Kconfig b/drivers/video/Kconfig
index 9f4c58f..523cddb 100644
--- a/drivers/video/Kconfig
+++ b/drivers/video/Kconfig
@@ -2486,2 +2486,3 @@ config FB_SIMPLE
 
+source "drivers/video/fbtft/Kconfig"
 source "drivers/video/omap/Kconfig"
diff --git a/drivers/video/Makefile b/drivers/video/Makefile
index 032f386..ede4965 100644
--- a/drivers/video/Makefile
+++ b/drivers/video/Makefile
@@ -143,2 +143,3 @@ obj-$(CONFIG_FB_SH_MOBILE_MERAM)  += sh_mobile_meram.o
 obj-$(CONFIG_FB_SH_MOBILE_LCDC)	  += sh_mobile_lcdcfb.o
+obj-$(CONFIG_FB_TFT)              += fbtft/
 obj-$(CONFIG_FB_OMAP)             += omap/
